src/v/serde/serde.h

```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
 template<typename T>
 void write(iobuf&, T);
 
+template<class R, class P>
+int64_t checked_duration_cast_to_nanoseconds(
+  const std::chrono::duration<R, P>& duration) {
+    static_assert(
+      __has_builtin(__builtin_mul_overflow),
+      "__builtin_mul_overflow not supported.");
+    using nano_period = std::chrono::nanoseconds::period;
+    using nano_rep = typename std::chrono::nanoseconds::rep;
+    // This is how duration_cast determines the output type.
+    using output_type = typename std::common_type<R, nano_rep, intmax_t>::type;
+    using ratio = std::ratio_divide<P, nano_period>;
+
+    static_assert(
+      std::is_same_v<output_type, nano_rep>,
+      "Output type is not the same rep as std::chrono::nanoseconds");
+
+    // Extra check to ensure the output type is same as the underlying type
+    // supported in lib[std]c++.
+    static_assert(
+      std::is_same_v<
+        output_type,
+        int64_t> || std::is_same_v<output_type, long long>,
+      "Output type not in supported integer types.");
+
+    constexpr auto ratio_num = static_cast<output_type>(ratio::num);
+    const auto dur = static_cast<output_type>(duration.count());
+    output_type mul;
+    if (unlikely(__builtin_mul_overflow(dur, ratio_num, &mul))) {
+        // Clamp the value.
+        // Log here to ensure that it is picked up by duck tape. Ideally this
+        // should never happen, but in case it does, we have a log trail.
+        //
+        // If you are here because your ducktape test failed with this BadLogLine
+        // check, it means that you serialized a duration type that caused an
+        // overflow when casting to nanoseconds. Clamp your duration to
+        // nanoseconds::max().
+        using input_type = typename std::chrono::duration<R, P>;
+        vlog(
+          serde_log.error,
+          "Overflow or underflow detected when casting to nanoseconds, "
+          "clamping to a limit. Input: {}  type: {}",
+          duration.count(),
+          type_str<input_type>());
+        return duration.count() < 0 ? std::chrono::nanoseconds::min().count()
+                                    : std::chrono::nanoseconds::max().count();
+    }
+    // No overflow/underflow detected, we are safe to cast.
+    return std::chrono::duration_cast<std::chrono::nanoseconds>(duration)
+      .count();
+}
+
 template<typename T>
 void write(iobuf& out, T t) {
     using Type = std::decay_t<T>;
+    static_assert(
+      !is_chrono_time_point<Type>,
+      "Time point serialization is risky and can have unintended "
+      "consequences. Check with RedPanda team before fixing this.");
```

Comment:
Capitalization: "Redpanda". I'm sorry for this nitpicking, [this](https://vectorizedio.atlassian.net/wiki/spaces/DOC/pages/92635182/Redpanda+Documentation+Style+Guide#Redpanda-specific-word-list) is still fresh on my memory :)

Reply:
haha nice catch.. done.